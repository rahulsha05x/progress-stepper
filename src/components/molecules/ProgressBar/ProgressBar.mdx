
---
React Progress Tracker
---

import { Playground } from 'docz';
import ProgressBar from './ProgressBar';
import './ProgressBar.scss';

# Progress Tracker

<Playground>
    <ProgressBar
        steps={[{"id": 1},{ "id": 2 },{ "id": 3},{ "id": 4}]}
        activeStep={5}
        orientation="horizontal"
        color="#4B0082"
        squared={false}
    />
</Playground>