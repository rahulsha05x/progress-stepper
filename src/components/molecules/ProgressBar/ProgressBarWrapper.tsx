import React from 'react';

export interface StepData {
  text?: string | object;
  status?: string;
  id: number;
  className: string;
  progress?: number;
}
interface Props {
  steps: StepData[];
  activeStep: number;
  className: string;
  orientation?: string;
  color?: string;
  content?: any;
}

export const getStatus = (activeStep: number, index: number) => {
  let stepClass = 'inactive';
  const completed = activeStep > index + 1;
  if (completed) {
    stepClass = 'is-completed';
  }
  if (activeStep === index + 1) {
    stepClass = 'is-active';
  }
  return stepClass;
};
const ProgressBarWrapper: React.FC<any> = ({
  steps,
  activeStep,
  align = 'center',
  orientation = 'horizontal',
  color,
  content,
  className = '',
  children,
}) => {
  let currentItemStatus: string;
  const renderChildren = () => {
    return React.Children.map(children, (child, index) => {
      currentItemStatus = getStatus(activeStep, index);
      return React.cloneElement(child, {
        currentitemstatus: currentItemStatus,
        itemindex: index,
      });
    });
  };
  const renderChild = renderChildren();
  return (
    <ul
      data-testid="progressBar-wrapper"
      className={`progress-tracker --${orientation} ${className}`}
    >
      {renderChild}
    </ul>
  );
};

export default ProgressBarWrapper;
