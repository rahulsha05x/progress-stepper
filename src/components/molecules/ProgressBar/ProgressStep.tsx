import React from 'react';
import Line from '../../atoms/Line/Line';
import Circle from '../../atoms/Circle/CircleIcon';
import TextWrapper from '../TextWrapper/TextWrapper';
import Step from '../StepperItem/Step';
import IconWrapper from '../IconContainer/IconWrapper';
interface Props {
  activeStep: number;
  className?: string;
  color?: string;
  icon?: any;
  completed?: boolean;
  text?: any;
  key?: string | number;
  progress?: number;
  orientation?: string;
  squared?: boolean;
  active?: boolean;
}

function renderProgressText(text: any) {
  if (typeof text === 'string') {
    return <div>{text}</div>;
  }

  if (React.isValidElement(text)) {
    return text;
  }
}

const ProgressStep: React.FC<Props> = ({
  text,
  activeStep,
  color = '#ccc',
  icon,
  completed = false,
  className = '',
  progress = 0,
  orientation = 'horizontal',
  squared = false,
  active = false,
}) => {
  console.log('Completed', completed);
  return (
    <Step
      color={color}
      completed={completed}
      content={icon}
      className={className}
    >
      <IconWrapper>
        <Circle
          active={active}
          completed={completed}
          icon={icon}
          color={color}
          squared={squared}
        />
        <Line
          active={active}
          completed={completed}
          progress={progress}
          color={color}
        />
      </IconWrapper>
      <TextWrapper>{renderProgressText(text)}</TextWrapper>
    </Step>
  );
};
ProgressStep.defaultProps = {
  completed: false,
};
export default ProgressStep;
