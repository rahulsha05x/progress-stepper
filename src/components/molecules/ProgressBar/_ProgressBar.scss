.ProgressBar {
  .hide {
    display: none;
  }
  &__Mobile {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    margin-left: 1.8rem;
    justify-content: space-evenly;
    min-height: 10rem;
    &__Text {
      font-weight: 600;
      color: rgba(0, 0, 0, 0.87);
      display: none;
      @include breakpoint-max(m-l) {
        display: inline-block;
      }
    }
    &__Status {
      color: silver;
    }
    &__Button {
      display: none;
      color: #fff;
      width: 100px;
      height: 50px;

      background-color: $blue-1;
      @include breakpoint-max(xl) {
        display: block;
      }
    }
  }

  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  border: 1px solid $light-gray;
  min-height: 95px;
  &__Sub {
    color: #4b4444;
    margin-left: 5.3rem;
    display: flex;
    justify-content: space-between;
    flex-wrap: nowrap;
    .fa {
      margin-right: 2rem;
    }
    @include breakpoint-max(m-l) {
      display: none;
    }
  }
  &__Main {
    display: flex;
    flex-wrap: nowrap;
    justify-content: flex-start;
    flex-basis: 20%;
    padding: 0 20px;
    @include breakpoint-max(m-l) {
      display: none;
    }
    &__StatusIndicator {
      background-color: $blue-1;
      border: 1px solid $blue-1;
      border-radius: 50%;
      height: 30px;
      width: 30px;
      align-self: center;
      margin-right: 33px;
    }
  }
  &__Steps {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding: 0;
    @include breakpoint-max(m-l) {
      flex-direction: row;
      position: relative;
    }
  }
}
