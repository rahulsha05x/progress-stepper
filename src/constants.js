export const stepsWithoutText = [
    {
      id: 1,
      status: 'Pending'
    },
    {
      id: 2,
      status: 'Pending'
    },
    {
      id: 3,
      status: 'Pending'
    },
    {
      id: 4,
      status: 'Pending'
    },
];

export const stepsWithText = [
    {
      id: 1,
      text: "Register your account",
      status: 'Pending'
    },
    {
      id: 2,
      text: 'Complete your wellness assessment', 
      status: 'Pending'
    },
    {
      id: 3,
      text: 'Link your account', 
      status: 'Pending'
    },
    {
      id: 4,
      text: 'Schedule your first call', 
      status: 'Pending'
    },
]