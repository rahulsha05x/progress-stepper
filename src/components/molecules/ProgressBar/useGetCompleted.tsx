import { StepData } from './ProgressBar';

const useGetCompleted = (steps: StepData[], completeFlag: string) => {
  const numberOfSteps = steps.length;
  const completedTask: StepData[] = steps.filter(
    (step) => step.status === completeFlag
  );
  const completedLen: number = completedTask.length;
  return {
    numberOfSteps,
    completedLen,
  };
};

export default useGetCompleted;
