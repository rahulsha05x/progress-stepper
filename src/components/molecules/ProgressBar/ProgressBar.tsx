import React from 'react';
import ProgressStep from './ProgressStep';
import ProgressBarWrapper from './ProgressBarWrapper';

export interface StepData {
    text?: string | object;
    status?: string;
    id: number;
    progress?: number;
    color?: string;
    icon?: string | object;
}
interface Props {
    steps: StepData[];
    activeStep: number;
    orientation?: string;
    color?: string;
    icon?: any;
    squared?: boolean;
}

const NewProgressBar: React.FC<Props> = ({
    steps = [],
    activeStep = 1,
    orientation = "horizontal",
    color,
    icon = <i className="fa fa-check"></i>,
    squared = false
}) => {
    return (
        <div>
            <ProgressBarWrapper activeStep={activeStep} orientation={orientation}>
                {steps.map((step: StepData, index: number) => {
                    let stepClass = '';
                    const completed = activeStep > index + 1;
                    if (completed) {
                        stepClass = 'is-completed';
                    }
                    if (activeStep === index + 1) {
                        stepClass = 'is-active';
                    }

                    return (
                        <ProgressStep
                            key={step.id || index}
                            {...step}
                            className={stepClass}
                            activeStep={activeStep}
                            completed={completed}
                            active={activeStep === index + 1}
                            color={step.color || color}
                            icon={step.icon || icon}
                            orientation={orientation}
                            squared={squared}
                        />
                    );
                })}
            </ProgressBarWrapper>
        </div>
    );
};

export default NewProgressBar;
