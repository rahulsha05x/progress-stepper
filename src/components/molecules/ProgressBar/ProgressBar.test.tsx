import React from 'react';
import { render, screen } from '@testing-library/react';
import ProgressBar from './ProgressBar';
import { stepsWithText } from '../../../constants';

describe('ProgressBarWrapper', () => {
  test('Renders ProgressBar wrapper', () => {
    const children = <li>Child example</li>;
    render(
      <ProgressBar steps={stepsWithText} activeStep={4}>
        {children}
      </ProgressBar>
    );

    expect(screen.getByTestId('progress-bar')).toMatchSnapshot();
  });
});
