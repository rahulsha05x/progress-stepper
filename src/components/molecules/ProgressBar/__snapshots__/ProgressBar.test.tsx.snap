// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarWrapper Renders ProgressBar wrapper 1`] = `
<div
  data-testid="progress-bar"
>
  <ul
    class="progress-tracker --horizontal "
    data-testid="progressBar-wrapper"
  >
    <li
      class="progress-step is-completed"
      data-testid="step"
    >
      <div
        class="progress-marker "
        data-testid="icon-wrapper"
      >
        <div
          class="circle "
          data-testid="circle"
          style="background: rgb(204, 204, 204); border-radius: 50%; border-color: #ccc;"
        >
          <i
            class="fa fa-check"
          />
        </div>
        <div
          class="line "
          data-testid="line"
          style="background-color: rgb(204, 204, 204);"
        />
      </div>
      <div
        class="progress-text "
      >
        <div>
          Step 1
        </div>
      </div>
    </li>
    <li
      class="progress-step is-completed"
      data-testid="step"
    >
      <div
        class="progress-marker "
        data-testid="icon-wrapper"
      >
        <div
          class="circle "
          data-testid="circle"
          style="background: rgb(204, 204, 204); border-radius: 50%; border-color: #ccc;"
        >
          <i
            class="fa fa-check"
          />
        </div>
        <div
          class="line "
          data-testid="line"
          style="background-color: rgb(204, 204, 204);"
        />
      </div>
      <div
        class="progress-text "
      >
        <div>
          Step 2
        </div>
      </div>
    </li>
    <li
      class="progress-step is-completed"
      data-testid="step"
    >
      <div
        class="progress-marker "
        data-testid="icon-wrapper"
      >
        <div
          class="circle "
          data-testid="circle"
          style="background: rgb(204, 204, 204); border-radius: 50%; border-color: #ccc;"
        >
          <i
            class="fa fa-check"
          />
        </div>
        <div
          class="line "
          data-testid="line"
          style="background-color: rgb(204, 204, 204);"
        />
      </div>
      <div
        class="progress-text "
      >
        <div>
          Step 3
        </div>
      </div>
    </li>
    <li
      class="progress-step is-active"
      data-testid="step"
    >
      <div
        class="progress-marker "
        data-testid="icon-wrapper"
      >
        <div
          class="circle "
          data-testid="circle"
          style="background: none; border-radius: 50%; border-color: #ccc;"
        />
        <div
          class="line "
          data-testid="line"
          style="background-color: rgb(204, 204, 204);"
        />
      </div>
      <div
        class="progress-text "
      >
        <div>
          Step 4
        </div>
      </div>
    </li>
  </ul>
</div>
`;
